src/Features/Core/Portable/CodeRefactorings/RefactoringHelpersService.cs

```diff
@@ -0,0 +1,89 @@
+
+        public Task<TSyntaxNode> TryGetSelectedNodeAsync<TSyntaxNode>(
+            Document document, TextSpan selection, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
+        {
+            return TryGetSelectedNodeAsync<TSyntaxNode>(document, selection, Functions<SyntaxNode>.Identity, cancellationToken);
+        }
+
+        public async Task<TSyntaxNode> TryGetSelectedNodeAsync<TSyntaxNode>(
+            Document document, TextSpan selection, Func<SyntaxNode, SyntaxNode> extractNode, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
+        {
+            return await TryGetSelectedNodeAsync(document, selection, extractNode, n => n is TSyntaxNode, cancellationToken).ConfigureAwait(false) as TSyntaxNode;
+        }
+
+        public async Task<SyntaxNode> TryGetSelectedNodeAsync(Document document, TextSpan selection, Func<SyntaxNode, SyntaxNode> extractNode, Predicate<SyntaxNode> predicate, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var selectionStripped = await CodeRefactoringHelpers.GetStrippedTextSpan(document, selection, cancellationToken).ConfigureAwait(false);
+
+            // Handle selections
+            var node = root.FindNode(selectionStripped, getInnermostNodeForTie: true);
+            SyntaxNode prevNode;
+            do
+            {
+                var extrNode = extractNode(node);
+                if (extrNode != default && predicate(extrNode))
+                {
+                    return extrNode;
+                }
+
+                prevNode = node;
+                node = node.Parent;
+
+            }
+            while (node != null && prevNode.FullWidth() == node.FullWidth());
+
+            // only consider what is direct selection touching when selection is empty 
+            // prevents `[|C|] methodName(){}` from registering as relevant for method Node
+            if (!selection.IsEmpty)
+            {
+                return default;
+            }
+
+            var tokenToLeft = await root.SyntaxTree.GetTouchingTokenToLeftAsync(selectionStripped.Start, cancellationToken).ConfigureAwait(false);
+            var leftNode = tokenToLeft.Parent;
+            do
+            {
+                // either touches a Token which parent is `TSyntaxNode` or is whose ancestor's span ends on selection
+                var extrNode = extractNode(leftNode);
+                if (extrNode != default && predicate(extrNode))
+                {
+                    return extrNode;
+                }
+
+                leftNode = leftNode?.Parent;
+            }
+            while (leftNode != null && leftNode.Span.End == selection.Start);
+
+            var tokenToRight = await root.SyntaxTree.GetTouchingTokenToRightOrInAsync(selectionStripped.Start, cancellationToken).ConfigureAwait(false);
```

Comment:
as per the other review, i dont' think we need these specific helpers that are only called from this single place.  just do the FindToken and check teh spans.

Reply:
As said in the other review, that doesn't work. I could move the helpers to local function here but we need their functionality. FindToken won't find token `void C(){}[||]` which we need to climb the right edge of MethodNode. The same situation is with GetTouchingToken that only gives you tokens along one edge (prefers left edge).

Reply:
i'm not sure i understand.  I'm basically saying that GetTouchingTokenToRightOrInAsync can just be moved to this file.  It's not needed as a general facility as a SyntaxTree extension.

Reply:
Ah, I see. Sure let's do that. I thought you wanted to get rid of them and use some simpler method.